pkg/controller/metrics/metricscontroller.go

```diff
@@ -0,0 +1,264 @@
+	Instance string `json:"instance"`
+	Job      string `json:"job"`
+}
+type Result struct {
+	Metric Metric        `json:"metric"`
+	Value  []interface{} `json:"value"`
+}
+type Data struct {
+	ResultType string   `json:"resultType"`
+	Result     []Result `json:"result"`
+}
+
+// latest+instant metrics structs end
+
+// latest+instant metrics structs begin
+type RangeMetricReponseFromPrometheus struct {
+	Status string    `json:"status"`
+	Data   RangeData `json:"data"`
+}
+type RangeMetric struct {
+	Name     string `json:"__name__"`
+	Device   string `json:"device"`
+	Instance string `json:"instance"`
+	Job      string `json:"job"`
+}
+type RangeResult struct {
+	Metric RangeMetric     `json:"metric"`
+	Values [][]interface{} `json:"values"`
+}
+type RangeData struct {
+	ResultType string        `json:"resultType"`
+	Result     []RangeResult `json:"result"`
+}
+
+// latest+instant metrics structs end
+
+func (c *controller) GetLatestMetrics(opt *pb.GetMetricsOpts) (*[]model.MetricSpec, error) {
+
+	var metrics []model.MetricSpec
+	// make a call to Prometheus, convert the response to our format, return
+	response, err := http.Get("http://localhost:9090/api/v1/query?query=" + opt.MetricName)
+	if err != nil {
+		log.Infof("The HTTP query request failed with error %s\n", err)
+	} else {
+		data, _ := ioutil.ReadAll(response.Body)
+		log.Info(string(data))
+
+		// unmarshal the JSON response into a struct (generated using the JSON, using this https://mholt.github.io/json-to-go/
+		var fv InstantMetricReponseFromPrometheus
+		err0 := json.Unmarshal(data, &fv)
+		log.Error(err0)
+
+		metrics := make([]model.MetricSpec, len(fv.Data.Result))
+
+		// now convert to our repsonse struct, so we can marshal it and send out the JSON
+		for i, res := range fv.Data.Result {
+			metrics[i].InstanceID = res.Metric.Instance + res.Metric.Device
+			metrics[i].Name = res.Metric.Name
+			metrics[i].MetricValues = make([]model.Metric, len(res.Value))
+			for j, v := range res.Value {
+				switch v.(type) {
+				case string:
+					metrics[i].MetricValues[j].Value, _ = strconv.ParseFloat(v.(string), 64)
+				case float64:
+					secs := int64(v.(float64))
+					metrics[i].MetricValues[j].Timestamp = secs
+				/*case []interface{}:
+				for i, u := range vv {
+					fmt.Println(i, u)
+				}*/
+				default:
+					log.Info(v, "is of a type I don't know how to handle")
+				}
+				/*metrics[i].CollectedAt = t
+				metrics[i].Value = v*/
+			}
+		}
+
+		log.Info("metrics struct is ", metrics)
+		bArr, _ := json.Marshal(metrics)
+		log.Info("metrics response json is ", string(bArr))
+
+		if err != nil {
+			log.Error(err)
+		}
+
+	}
+	return &metrics, err
+}
+
+func (c *controller) GetInstantMetrics(opt *pb.GetMetricsOpts) (*[]model.MetricSpec, error) {
+
+	var metrics []model.MetricSpec
+	// make a call to Prometheus, convert the response to our format, return
+	response, err := http.Get("http://localhost:9090/api/v1/query?query=" + opt.MetricName + "&time=" + opt.StartTime)
```

Comment:
Please remove the hard coding.
The endpoint addr could be a config and the rest could be specific to the method.

Reply:
https://github.com/opensds/opensds/issues/691